### ISSUE TITLE
UI freeze on startup due org.eclipse.xtext.ui.shared.internal.ListenerRegistrar.initialize

### ISSUE THREAD

**Issue by iloveeclipse** [eclipse/xtext-eclipse#1129]
Just got a minute long freeze on startup in the 4.12 IDE / XText 2.18.

Looks like XText itself has a (timed out) lock on loading between JavaReconciler thread and "Worker-47: Build" thread, both thrying to load XText classes in parallel.

`org.eclipse.xtext.ui.shared.internal.Activator.start` locks OSGI activation for XText, but waits on the workspace lock that is held by the `"Worker-47: Build"` thread, where `org.eclipse.xtext.ui.ecore.Activator.start` tries to activate xtext.ui.shared.

Looks like `workspaceLockAccess.isWorkspaceLockedByCurrentThread(workspace)` says "no", the workspace is not locked by current thread, so it tries to run `workspace.run(runnable, null, IWorkspace.AVOID_UPDATE, null)`.

This will NOT interrupt autobuild because the given rule is null, instead it will lock if the autobuild is already running via `org.eclipse.core.internal.resources.WorkManager.checkIn(ISchedulingRule, IProgressMonitor)`.

I believe that by specifying the rule in `workspace.run(runnable, null, IWorkspace.AVOID_UPDATE, null)` we could trigger autobuild cancellation so that the lock does not occur. However cancelling autobuild is also probably not ideal, and selecting a matching rule is always hard (if possible at all).

Therefore ideally the code in `org.eclipse.xtext.ui.shared.internal.ListenerRegistrar.initialize()` would run this block in a job:
`
if (builderState.isEmpty()) {		buildManager.scheduleBuildIfNecessary(Arrays.asList(workspace.getRoot().getProjects()), IBuildFlag.RECOVERY_BUILD);
}
`
This would fix the issue for sure.

[xtext_freeze_on_startup_short_version.log](https://github.com/eclipse/xtext-eclipse/files/3368652/xtext_freeze_on_startup_short_version.log)

**Comment by cdietrich** [comment]
cc @szarnekow

**Comment by trancexpress** [comment]
My patch suggestion is:

```
diff --git a/org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/internal/ListenerRegistrar.java b/org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/internal/ListenerRegistrar.java
index c62def94a..4722331d7 100644
--- a/org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/internal/ListenerRegistrar.java
+++ b/org.eclipse.xtext.ui.shared/src/org/eclipse/xtext/ui/shared/internal/ListenerRegistrar.java
@@ -10,6 +10,11 @@ package org.eclipse.xtext.ui.shared.internal;
 import java.util.Arrays;
 
 import org.eclipse.core.resources.IWorkspace;
+import org.eclipse.core.resources.WorkspaceJob;
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.IStatus;
+import org.eclipse.core.runtime.Status;
 import org.eclipse.xtext.builder.builderState.IBuilderState;
 import org.eclipse.xtext.builder.impl.BuildScheduler;
 import org.eclipse.xtext.builder.impl.IBuildFlag;
@@ -41,9 +46,25 @@ public class ListenerRegistrar implements IEagerContribution {
 		// after the fact, the project-open event is already done. Thus no build is triggered for the newly
 		// opened project and the index state is corrupt.
 		// thus we trigger a recovery build here
-		if (builderState.isEmpty()) {
-			buildManager.scheduleBuildIfNecessary(Arrays.asList(workspace.getRoot().getProjects()), IBuildFlag.RECOVERY_BUILD);
+		class RecoveryBuildTrigger extends WorkspaceJob {
+
+			public RecoveryBuildTrigger() {
+				super("Schedule xtext recovery build on start-up");
+			}
+
+			@Override
+			public IStatus runInWorkspace(IProgressMonitor monitor) throws CoreException {
+				if (monitor.isCanceled()) {
+					return Status.CANCEL_STATUS;
+				}
+				if (builderState.isEmpty()) {
+					buildManager.scheduleBuildIfNecessary(Arrays.asList(workspace.getRoot().getProjects()), IBuildFlag.RECOVERY_BUILD);
+				}
+				return Status.OK_STATUS;
+			}
 		}
+		RecoveryBuildTrigger recoveryBuildTrigger = new RecoveryBuildTrigger();
+		recoveryBuildTrigger.schedule();
 	}
 
 	@Override
```

This should avoid trying to hold the workspace log during this stack trace:

```
"org.eclipse.jdt.internal.ui.text.JavaReconciler" #31 daemon prio=1 os_prio=0 tid=0x00007ffff2506000 nid=0x3ae5 in Object.wait() [0x00007fff7110d000]
   java.lang.Thread.State: TIMED_WAITING (on object monitor)
	at java.lang.Object.wait(Native Method)
	at org.eclipse.core.internal.jobs.Semaphore.acquire(Semaphore.java:44)
	- locked <0x00000005daede320> (a org.eclipse.core.internal.jobs.Semaphore)
	at org.eclipse.core.internal.jobs.OrderedLock.doAcquire(OrderedLock.java:173)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:109)
	at org.eclipse.core.internal.jobs.OrderedLock.acquire(OrderedLock.java:85)
	at org.eclipse.core.internal.resources.WorkManager.endUnprotected(WorkManager.java:190)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2307)
	at org.eclipse.core.internal.resources.Workspace.run(Workspace.java:2322)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.load(AbstractBuilderState.java:91)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.ensureLoaded(AbstractBuilderState.java:106)
	at org.eclipse.xtext.builder.builderState.AbstractBuilderState.isEmpty(AbstractBuilderState.java:253)
	at org.eclipse.xtext.ui.shared.internal.ListenerRegistrar.initialize(ListenerRegistrar.java:44)
	at org.eclipse.xtext.ui.shared.internal.EagerContributionInitializer.initialize(EagerContributionInitializer.java:31)
	at org.eclipse.xtext.ui.shared.internal.Activator.start(Activator.java:214)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:842)
	at org.eclipse.osgi.internal.framework.BundleContextImpl$3.run(BundleContextImpl.java:1)
	at java.security.AccessController.doPrivileged(Native Method)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.startActivator(BundleContextImpl.java:834)
	at org.eclipse.osgi.internal.framework.BundleContextImpl.start(BundleContextImpl.java:791)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.startWorker0(EquinoxBundle.java:1013)
	at org.eclipse.osgi.internal.framework.EquinoxBundle$EquinoxModule.startWorker(EquinoxBundle.java:365)
	at org.eclipse.osgi.container.Module.doStart(Module.java:598)
	at org.eclipse.osgi.container.Module.start(Module.java:462)
	at org.eclipse.osgi.framework.util.SecureAction.start(SecureAction.java:493)
	at org.eclipse.osgi.internal.hooks.EclipseLazyStarter.postFindLocalClass(EclipseLazyStarter.java:117)
	at org.eclipse.osgi.internal.loader.classpath.ClasspathManager.findLocalClass(ClasspathManager.java:570)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.findLocalClass(ModuleClassLoader.java:330)
	at org.eclipse.osgi.internal.loader.BundleLoader.findLocalClass(BundleLoader.java:396)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClassInternal(BundleLoader.java:474)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:423)
	at org.eclipse.osgi.internal.loader.BundleLoader.findClass(BundleLoader.java:415)
	at org.eclipse.osgi.internal.loader.ModuleClassLoader.loadClass(ModuleClassLoader.java:155)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at org.eclipse.osgi.internal.framework.EquinoxBundle.loadClass(EquinoxBundle.java:620)
	at org.eclipse.core.internal.registry.osgi.RegistryStrategyOSGI.createExecutableExtension(RegistryStrategyOSGI.java:196)
	at org.eclipse.core.internal.registry.ExtensionRegistry.createExecutableExtension(ExtensionRegistry.java:934)
	at org.eclipse.core.internal.registry.ConfigurationElement.createExecutableExtension(ConfigurationElement.java:246)
	at org.eclipse.core.internal.registry.ConfigurationElementHandle.createExecutableExtension(ConfigurationElementHandle.java:63)
	at org.eclipse.jdt.internal.core.JavaModelManager$CompilationParticipants$1.run(JavaModelManager.java:466)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.core.JavaModelManager$CompilationParticipants.getCompilationParticipants(JavaModelManager.java:459)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.notifyParticipants(ReconcileWorkingCopyOperation.java:241)
	at org.eclipse.jdt.internal.core.ReconcileWorkingCopyOperation.executeOperation(ReconcileWorkingCopyOperation.java:99)
	at org.eclipse.jdt.internal.core.JavaModelOperation.run(JavaModelOperation.java:736)
	at org.eclipse.jdt.internal.core.JavaModelOperation.runOperation(JavaModelOperation.java:802)
	at org.eclipse.jdt.internal.core.CompilationUnit.reconcile(CompilationUnit.java:1322)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:131)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.access$0(JavaReconcilingStrategy.java:113)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy$1.run(JavaReconcilingStrategy.java:93)
	at org.eclipse.core.runtime.SafeRunner.run(SafeRunner.java:45)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.reconcile(JavaReconcilingStrategy.java:90)
	at org.eclipse.jdt.internal.ui.text.java.JavaReconcilingStrategy.initialReconcile(JavaReconcilingStrategy.java:188)
	at org.eclipse.jdt.internal.ui.text.CompositeReconcilingStrategy.initialReconcile(CompositeReconcilingStrategy.java:122)
	at org.eclipse.jdt.internal.ui.text.JavaCompositeReconcilingStrategy.initialReconcile(JavaCompositeReconcilingStrategy.java:139)
	at org.eclipse.jface.text.reconciler.MonoReconciler.initialProcess(MonoReconciler.java:98)
	at org.eclipse.jdt.internal.ui.text.JavaReconciler.initialProcess(JavaReconciler.java:423)
	- locked <0x00000005d0805190> (a java.lang.Object)
	at org.eclipse.jface.text.reconciler.AbstractReconciler$BackgroundThread.run(AbstractReconciler.java:177)
```

And so avoid the issue. Christian, Sebastian, any thoughts on this?

**Comment by szarnekow** [comment]
The patch looks good to me. 
I'm a little concerned though that some other call graph will call `ClusteringBuilderState.ensureLoaded` and we'll run into the same problem. But this is for now only a gut feeling.

**Comment by szarnekow** [comment]
In other words: A PR for this is appreciated.

**Comment by trancexpress** [comment]
Would it be OK to set PRs not only for master but also for the xtext 2.18 maintenance branch?

**Comment by szarnekow** [comment]
As of now we do not have plans to release a 2.18.x

**Comment by cdietrich** [comment]
Should be fixed in 2.20
